src/optimizer/optimizer.cpp

```diff
@@ -336,5 +339,30 @@ unique_ptr<planner::AbstractPlan> Optimizer::ChooseBestPlan(
   LOG_TRACE("Finish Choosing best plan for group %d", id);
   return plan;
 }
+
+void Optimizer::ExecuteTaskStack(
+    std::unique_ptr<OptimizerTaskStack> &task_stack, const Group *root_group,
+    std::shared_ptr<OptimizeContext> root_context) {
+  auto &timer = root_context->timer;
+  const auto &required_props = root_context->required_prop;
+  const auto timeout_limit = root_context->timeout_limit;
+  // Iterate through the task stack
+  while (!task_stack->Empty()) {
+    // Check to see if we have at least one plan, and if we have exceeded our
+    // timeout limit
+    if (timer.GetDuration() >= timeout_limit &&
+        root_group->HasExpressions(required_props)) {
+      throw OptimizerException("Optimizer task execution duration " +
+                               std::to_string(timer.GetDuration()) +
+                               " exceeds timeout limit " +
+                               std::to_string(timeout_limit));
+    }
+    timer.Start();
```

Comment:
Looks like we reset the timer for each task? I think we should start the timer once for each query.

Reply:
we aren't resetting the timer here. we are simply stopping it after a given task execution and checking if the total timer duration exceeds the limit.

Reply:
I see what you mean. But this does not seem to be measuring the actual execution time of the optimizer because some part of the execution hasn't been counted. Could we only call `Start()` once and after calling `Stop()` to capture the total duration, we call `Reset()` to zero `elapse_`?

Reply:
I think I am confused sorry - wouldn't this make it so that we can't stop in the middle of task execution if we went over the time limit?